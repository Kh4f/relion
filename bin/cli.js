#!/usr/bin/env node

import ryly from '../src/index.js';
import cmdParser from '../src/commands.js';

ryly(cmdParser.argv).catch((err) => {
	console.error(err);
	process.exit(1);
});
