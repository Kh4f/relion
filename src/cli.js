#!/usr/bin/env node

import relion from './index.js'
import cmdParser from './commands.js'

relion(cmdParser.argv).catch((err) => {
	console.error(err)
	process.exit(1)
})