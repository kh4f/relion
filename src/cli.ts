import { resolve } from 'node:path'
import { pathToFileURL } from 'node:url'
import relion from './index.js'
import type { UserConfig } from '@/types'
import { cli } from 'cleye'

if (import.meta.main) await runCli()

export async function runCli(config?: UserConfig, argvs?: string[]) {
	const argv = cli({
		name: 'relion',
		flags: {
			bump: {
				alias: 'b',
				type: Boolean,
				description: 'Bump the version',
				default: false,
			},
			changelog: {
				alias: 'l',
				type: Boolean,
				description: 'Generate a changelog',
				default: false,
			},
			commit: {
				alias: 'c',
				type: Boolean,
				description: 'Create a commit',
				default: false,
			},
			tag: {
				alias: 't',
				type: Boolean,
				description: 'Create a tag',
				default: false,
			},
			profile: {
				alias: 'p',
				type: String,
				description: 'Use config profile',
			},
			dryRun: {
				alias: 'd',
				type: Boolean,
				description: 'Run without making any changes',
				default: false,
			},
			latest: {
				alias: 'L',
				type: Boolean,
				description: 'Use the latest-release commit range in changelog',
				default: false,
			},
		},
	}, undefined, argvs)

	if (!config) {
		try {
			const configFileURL = pathToFileURL(resolve(process.cwd(), 'relion.config.ts')).href
			config = ((await import(configFileURL)) as { default: UserConfig }).default
		} catch (error) {
			throw new Error(`Error loading config: ${(error as Error).message}`)
		}
	}

	// argv may be undefined if --help was passed
	if (!(argv as ReturnType<typeof cli> | undefined)) return

	if (!argv.flags.bump) config.bump = false
	if (!argv.flags.changelog) config.changelog = false
	if (!argv.flags.commit) config.commit = false
	if (!argv.flags.tag) config.tag = false
	if (argv.flags.profile) config.profile = argv.flags.profile
	if (argv.flags.dryRun) config.dryRun = argv.flags.dryRun
	if (argv.flags.latest && config.changelog) {
		if (config.changelog === true) config.changelog = {}
		config.changelog.commitRange = 'latest-release'
	}

	return await relion(config)
}