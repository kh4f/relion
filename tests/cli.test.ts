import { describe, it, expect, vi } from 'vitest'
import { runCli } from '@/cli'
import relionProjectConfig from '../relion.config'
import relionTestConfig from './fixtures/relion.config.cli'

describe('runCli', () => {
	it('should output help message when `--help` is passed', async () => {
		const exitSpy = vi.spyOn(process, 'exit').mockImplementation(() => undefined as never)
		let log = ''
		const logSpy = vi.spyOn(console, 'log').mockImplementation((msg: string) => log += msg)
		await runCli('--help', {})
		expect(exitSpy).toHaveBeenCalledWith(0)
		expect(log).toMatchSnapshot()
		logSpy.mockRestore()
	})

	it('should load config from default path (relion.config.ts)', async () => {
		expect((await runCli('-f l'))?.inputConfig).toEqual(relionProjectConfig)
	})

	it('should load config from custom path', async () => {
		expect((await runCli('--config tests/fixtures/relion.config.cli.ts --dry'))?.inputConfig).toEqual(relionTestConfig)
	})
})

describe('config transformation via CLI args', () => {
	it('should apply CLI flags to implicit default profile', async () => {
		expect((await runCli('-f l', {}))?.inputConfig).toEqual({
			_default: { lifecycle: ['changelog'] },
		})
	})

	it('should apply CLI flags to explicitly specified default profile', async () => {
		expect((await runCli('-L', { profile: 'default', _default: {} }))?.inputConfig).toEqual({
			profile: 'default',
			_default: { lifecycle: 'all', changelog: { commitRange: 'latest-release' } },
		})
	})

	it('should apply CLI flags to specified custom profile', async () => {
		expect((await runCli('-f l -d', { profile: 'custom', _custom: { lifecycle: 'all' } }))?.inputConfig).toEqual({
			profile: 'custom',
			_custom: { lifecycle: ['changelog'], dryRun: true },
		})
	})

	it('should throw error when specified profile does not exist in config', async () => {
		await expect(() => runCli('', { profile: 'custom' })).rejects.toThrowError(`Profile 'custom' not found in config`)
	})

	it('should throw error when explicitly specified default profile is not defined', async () => {
		await expect(() => runCli('', { profile: 'default' })).rejects.toThrowError(`Profile 'default' not found in config`)
	})
})

describe('lifecycle flag parsing', () => {
	it('should set lifecycle to "all" when "-f all" is passed', async () => {
		expect((await runCli('-f all', {}))?.inputConfig._default).toEqual({ lifecycle: 'all' })
	})

	it('should map "-f l" to ["changelog"]', async () => {
		expect((await runCli('-f l', {}))?.inputConfig._default).toEqual({ lifecycle: ['changelog'] })
	})

	it('should map "-f bt" to ["bump","tag"]', async () => {
		expect((await runCli('-f bt', {}))?.inputConfig._default).toEqual({ lifecycle: ['bump', 'tag'] })
	})

	it('should parse composite alias "-f tmlb" and preserves order', async () => {
		expect((await runCli('-f tmlb', {}))?.inputConfig._default).toEqual({ lifecycle: ['tag', 'commit', 'changelog', 'bump'] })
	})

	it('should throw on invalid lifecycle alias', async () => {
		await expect(() => runCli('-f tclb', {})).rejects.toThrowError(`Invalid lifecycle step alias: 'c'`)
	})
})