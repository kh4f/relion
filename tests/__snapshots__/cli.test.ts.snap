// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runCli > should output help message when \`--help\` is passed 1`] = `
"relion

[1mUsage:[22m
  relion [flags...]

[1mFlags:[22m
  -c, --config <string>           Path to the config file (default: "relion.config.ts")
  -d, --dry                       Run without making any changes
  -h, --help                      Show help
  -L, --latest                    Use the latest-release commit range in changelog
  -f, --lifecycle <string>        Lifecycle steps to run in order ((b)ump, change(l)og, co(m)mit, (t)ag, or "all"). (default: "all")
  -p, --profile <string>          Use config profile
"
`;
